p2p/testing/protocolsession.go

```diff
@@ -0,0 +1,213 @@
+	simNode, ok := self.adapter.GetNode(trig.Peer)
+	if !ok {
+		return fmt.Errorf("trigger: peer %v does not exist (1- %v)", trig.Peer, len(self.IDs))
+	}
+	mockNode, ok := simNode.Services()[0].(*mockNode)
+	if !ok {
+		return fmt.Errorf("trigger: peer %v is not a mock", trig.Peer)
+	}
+
+	errc := make(chan error)
+
+	go func() {
+		errc <- mockNode.Trigger(&trig)
+	}()
+
+	t := trig.Timeout
+	if t == time.Duration(0) {
+		t = 1000 * time.Millisecond
+	}
+	select {
+	case err := <-errc:
+		return err
+	case <-time.After(t):
+		return fmt.Errorf("timout expecting %v to send to peer %v", trig.Msg, trig.Peer)
+	}
+}
+
+// expect checks an expectation
+func (self *ProtocolSession) expect(exp Expect) error {
+	if exp.Msg == nil {
+		return errors.New("no message to expect")
+	}
+	simNode, ok := self.adapter.GetNode(exp.Peer)
+	if !ok {
+		return fmt.Errorf("trigger: peer %v does not exist (1- %v)", exp.Peer, len(self.IDs))
+	}
+	mockNode, ok := simNode.Services()[0].(*mockNode)
+	if !ok {
+		return fmt.Errorf("trigger: peer %v is not a mock", exp.Peer)
+	}
+
+	errc := make(chan error)
+	go func() {
+		errc <- mockNode.Expect(&exp)
+	}()
+
+	t := exp.Timeout
+	if t == time.Duration(0) {
+		t = 2000 * time.Millisecond
+	}
+	select {
+	case err := <-errc:
+		return err
+	case <-time.After(t):
+		return fmt.Errorf("timout expecting %v sent to peer %v", exp.Msg, exp.Peer)
+	}
+}
+
+// TestExchanges tests a series of exchanges againsts the session
+func (self *ProtocolSession) TestExchanges(exchanges ...Exchange) error {
+	// launch all triggers of this exchanges
+
+	for i, e := range exchanges {
+		errc := make(chan error, 1)
+		wg := &sync.WaitGroup{}
+		for _, trig := range e.Triggers {
+			err := self.trigger(trig)
+			if err != nil {
+				errc <- err
+				break
+			}
+		}
+
+		// each expectation is spawned in separate go-routine
+		// expectations of an exchange are conjunctive but unordered, i.e.,
+		// only all of them arriving constitutes a pass
+		// each expectation is meant to be for a different peer, otherwise they are expected to panic
+		// testing of an exchange blocks until all expectations are decided
+		// an expectation is decided if
+		//  expected message arrives OR
+		// an unexpected message arrives (panic)
+		// times out on their individual timeout
+		for _, ex := range e.Expects {
+			wg.Add(1)
+			// expect msg spawned to separate go routine
+			go func(exp Expect) {
+				defer wg.Done()
+				err := self.expect(exp)
+				if err != nil {
+					errc <- err
+				}
+			}(ex)
+		}
+
+		// wait for all expectations
+		go func() {
+			wg.Wait()
+			close(errc)
+		}()
+
+		// time out globally or finish when all expectations satisfied
+		alarm := time.NewTimer(1000 * time.Millisecond)
+		select {
+
+		case err := <-errc:
+			if err != nil {
+				return fmt.Errorf("exchange failed with: %v", err)
+			}
+		case <-alarm.C:
+			return fmt.Errorf("exchange %v: '%v' timed out", i, e.Label)
+		}
```

Comment:
This select waits and returns a timeout if there is no error on errc, why is that?

Reply:
`errc` is closed on line 168 after waiting for all expects to finish, which would then trigger this select to pass through. If the timeout is hit, the expects just didn't finish in time.

That being said, I think this could be simpler:

```go
errc := make(chan error, len(e.Expects))

for _, ex := range e.Expects {
	go func(ex Expect) { errc <- self.expect(ex) }(ex)
}

timeout := time.After(time.Second)
for i := 0; i < len(e.Expects); i++ {
	select {
	case err := <-errc:
		// check err
	case <-timeout:
		// return timeout
	}
}
```